File: src/core/messenger/DownloadController.ts
/**
 * DownloadController.ts - org.telegram.messenger.DownloadController
 * Direct TypeScript translation of DrKLO/Telegram Android architecture
 */

import { Message } from '../../types';
import { NotificationCenter } from '../NotificationCenter';

export class DownloadController {
  public static readonly AUTODOWNLOAD_MASK_PHOTO = 1;
  public static readonly AUTODOWNLOAD_MASK_AUDIO = 2;
  public static readonly AUTODOWNLOAD_MASK_VIDEO = 4;
  public static readonly AUTODOWNLOAD_MASK_DOCUMENT = 8;

  private static instances = new Map<number, DownloadController>();
  private currentAccount: number;

  public static getInstance(account: number = 0): DownloadController {
    if (!DownloadController.instances.has(account)) {
      DownloadController.instances.set(account, new DownloadController(account));
    }
    return DownloadController.instances.get(account)!;
  }

  private constructor(account: number) {
    this.currentAccount = account;
  }

  public canDownloadMedia(type: number): boolean {
    const currentMask =
      DownloadController.AUTODOWNLOAD_MASK_PHOTO |
      DownloadController.AUTODOWNLOAD_MASK_AUDIO |
      DownloadController.AUTODOWNLOAD_MASK_DOCUMENT;
    return (currentMask & type) !== 0;
  }

  public checkAndDownloadMedia(message: Message): void {
    if (!message || !message.media) return;
    // Handled via FileLoader
  }
}

export const downloadController = DownloadController.getInstance();

File: TMessagesProj/src/main/java/org/telegram/messenger/DownloadController.java
/*
 * This is the source code of Telegram for Android v. 12.x.x.
 * It is licensed under GNU GPL v. 2 or later.
 * You should have received a copy of the license in this archive (see LICENSE).
 *
 * Copyright Nikolai Kudashov, 2013-2024.
 */

package org.telegram.messenger;

import org.telegram.tgnet.TLRPC;

public class DownloadController {
    public static final int AUTODOWNLOAD_MASK_PHOTO = 1;
    public static final int AUTODOWNLOAD_MASK_AUDIO = 2;
    public static final int AUTODOWNLOAD_MASK_VIDEO = 4;
    public static final int AUTODOWNLOAD_MASK_DOCUMENT = 8;

    private static volatile DownloadController[] Instance = new DownloadController[UserConfig.MAX_ACCOUNT_COUNT];

    public static DownloadController getInstance(int num) {
        DownloadController localInstance = Instance[num];
        if (localInstance == null) {
            synchronized (DownloadController.class) {
                localInstance = Instance[num];
                if (localInstance == null) {
                    Instance[num] = localInstance = new DownloadController(num);
                }
            }
        }
        return localInstance;
    }

    public DownloadController(int account) {}

    public static boolean canDownloadMedia(int type, int networkType) {
        int currentMask = AUTODOWNLOAD_MASK_PHOTO | AUTODOWNLOAD_MASK_AUDIO | AUTODOWNLOAD_MASK_DOCUMENT;
        return (currentMask & type) != 0;
    }

    public static void checkAndDownloadMedia(TLRPC.Message message) {
        if (message == null || message.media == null) return;
        if (message.media instanceof TLRPC.TL_messageMediaPhoto) {
            FileLoader.getInstance(UserConfig.selectedAccount).loadFile(
                ImageLocation.getForPhoto(FileLoader.getClosestPhotoSizeWithSize(message.media.photo.sizes, 1280), message.media.photo),
                message, "jpg", 1, 1
            );
        } else if (message.media instanceof TLRPC.TL_messageMediaDocument) {
            FileLoader.getInstance(UserConfig.selectedAccount).loadFile(
                message.media.document, message, 0, 1
            );
        }
    }
}
